doc/guide/sonata-io.md

```diff
@@ -0,0 +1,202 @@
+# Using Sonata IO
+
+## Overview
+
+The v1.0 Sonata System contains the following peripheral blocks for general purpose use:
+
+ - 5x SPI (2 dedicated and 3 muxable)
+ - 3x UART (1 dedicated and 2 muxable)
+ - 2x I2C (both muxable)
+ - GPIO (LEDs and Switches on the board plus GPIO for all headers other than
+   mikroBUS)
+ - PWM (6 muxable channels, 1 dedicated channel)
+ - ADC (6 dedicated channels)
+ - 1x USB Device (dedicated)
+
+Sonata contains a [pinmux][] which allows you to chose which block is connected to a particular physical IO pin.
+Each pin has its own unique selection of blocks it can be attached to.
+The 'muxable' in the list refers to the blocks that can connect to multiple pins.
+Some of the blocks are dedicated, they connect to specific IO, e.g. there is a dedicated SPI block for the Ethernet controller.
+
+## Pinmux Usage
+
+The pinmux has two kinds of selectors:
+
+- Output selectors - One per pin, chooses which block I/O outputs to that pin
+- Input selectors - One per block IO, chooses which pin inputs to that block IO
+
+A [pinmux driver][] is available to allow you to manipulate these selectors.
+Note there are defined constants for the selectors themselves but not the options within the selectors.
+E.g. PMOD0 pin 2 has output selector constant `OutputPin::pmod0_2` but a raw number is used to choose which block I/O is connected to.
+The [pinmux documentation][] provides the possible inputs and outputs for each selector.
+
+For example say you wished to connect the PMOD0 SPI to SPI block 1; there are 4 selectors to set.
+
+- PMOD 0 CS (pin 1), COPI (pin 2) and SCK (pin 4) output selectors
+- SPI Block 1 CIPO input selector
+
+The code below shows how to use the pinmux driver to do this:
+
+```c++
+#include <platform-pinmux.hh>
+
+...
+
+auto pinmux = SonataPinmux();
+pinmux->output_pin_select(SonataPinmux::OutputPin::pmod0_1, 2);
+pinmux->block_input_select(SonataPinmux::BlockInput::spi_1_cipo, 3);
+pinmux->output_pin_select(SonataPinmux::OutputPin::pmod0_2, 2);
+pinmux->output_pin_select(SonataPinmux::OutputPin::pmod0_4, 2);
```

Comment:
```suggestion
pinmux.output_pin_select(SonataPinmux::OutputPin::pmod0_1, 2);
pinmux.block_input_select(SonataPinmux::BlockInput::spi_1_cipo, 3);
pinmux.output_pin_select(SonataPinmux::OutputPin::pmod0_2, 2);
pinmux.output_pin_select(SonataPinmux::OutputPin::pmod0_4, 2);
```
Still not quite correct - `SonataPinmux` is now an object and not a pointer, so `->` usage should be replaced with `.`

Reply:
Ah whoops.

Maybe I should do something radical like actually compile my code snippet and check it works...

Reply:
Have demonstrated this working in a test program. May add that as an example (demonstrates using the pmod SPI) but that's a job for Monday!